[mlir][Interfaces][NFC] `ValueBoundsConstraintSet`: Add columns for constant values/dims

matthias-springer opened the pull request:
`ValueBoundsConstraintSet` maintains an internal constraint set (`IntegerRelation`), where every analyzed index-typed SSA value or dimension of a shaped type is represented with a dimension/symbol. Prior to this change, index-typed values with a statically known constant value and static shaped type dimensions were not added to the constraint set. Instead, `getExpr` directly returned an affine constant expression.

With this commit, dynamic and static values/dimension sizes are treated in the same way: in either case, a dimension/symbol is added to the constraint set. This is needed for a subsequent commit that adds support for branches.

nicolasvasilache commented on mlir/lib/Interfaces/ValueBoundsOpInterface.cpp:
```diff
@@ -107,25 +107,57 @@ AffineExpr ValueBoundsConstraintSet::getExpr(Value value,
+  auto getPosExpr = [&](int64_t pos) {
+    assert(pos >= 0 && pos < cstr.getNumDimAndSymbolVars() &&
+           "invalid position");
+    return pos < cstr.getNumDimVars()
```
This ternary condition feels like an implementation detail that appears everywhere.
Can we give it a proper name and hide away the impl ?

matthias-springer replied:
I have a WIP change that turns this into a member function (so that it can be called from various places). Can also be done now, I moved it into this PR.

matthias-springer committed:
[mlir][Interfaces][NFC] `ValueBoundsConstraintSet`: Add columns for constant values/dims

`ValueBoundsConstraintSet` maintains an internal constraint set (`IntegerRelation`), where every analyzed index-typed SSA value or dimension of a shaped type is represented with a dimension/symbol. Prior to this change, index-typed values with a statically known constant value and static shaped type dimensions were not added to the constraint set. Instead, `getExpr` directly returned an affine constrant expression.

With this commit, dynamic and static values/dimension sizes are treated in the same way: in either case, a dimension/symbol is added to the constraint set. This is needed for a subsequent commit that adds support for branches.

nicolasvasilache commented on mlir/lib/Interfaces/ValueBoundsOpInterface.cpp:
```diff
@@ -107,25 +107,47 @@ AffineExpr ValueBoundsConstraintSet::getExpr(Value value,
-    // Constant index value: return directly.
-    if (auto constInt = ::getConstantIntValue(value))
-      return builder.getAffineConstantExpr(*constInt);
+      constSize = shapedType.getDimSize(*dim);
```
Hmm, I feel this is tricky: this really depends about the context in which we query the analysis.
It is easy to create a varying shape (i.e. size depends on loop iv) and this will not be a symbol anymore.

Basically, if the analysis was properly rooted at an op / block, we could say that everything defined above is a symbol.

IOW, any such analysis is relative to its context: if your transform/analysis cares about information on say k innermost loops, you can "make" information about the outermost loops constant.

matthias-springer replied:
I don't fully understand the different between symbols and dimensions in `IntegerRelation`. The limitations that I am aware of:

- `getSliceBounds` can only be called on a dimensions, so the first "column" that I add to the constraint set (for which we want to compute LB/UB) is a dimension.
- `projectOut` works on both symbols and dimensions, so all other values are added as symbols.

Basically we have only one dimension and everything else is a symbol. (But the current implementation would make it easy to change this.) This seems to work well so far.

matthias-springer replied:
> any such analysis is relative to its context

On second thought, that's actually how it is implemented at the moment. We provide a `ValueBoundsConstraintSet::computeBound(Value)` API. That function will build a constraint set and add the given value as a dimension, everything else as a symbol. When computing a bound for a different value, we build a brand new constraint set, this time with the new value being the only dimension. So this analysis is actually relative to a context and rooted at a certain op.

Computing multiple bounds back-to-back is slow at the moment because we have to re-analyze the entire IR. If we improve this at some point in the future, we will have to take a closer look at what's a symbol and what's a dimension, as you suggested.

jpienaar reviewed (commented):
This makes sense to me (and if I follow the discussion resolves Nicholas's question). What is the performance impact of this? This seems to potentially result in a lot more constraints. Also is it possible to test?

matthias-springer commented:
> This makes sense to me (and if I follow the discussion resolves Nicholas's question).

Yes, I think so.


> What is the performance impact of this? This seems to potentially result in a lot more constraints. Also is it possible to test?

A few more constraints are added, but not many. E.g., when we have IR such as:
```mlir
%0 = tensor.insert %f into %t[%idx] : tensor<5xf32>
%1 = tensor.insert %f into %0[%idx] : tensor<5xf32>
%2 = tensor.insert %f into %1[%idx] : tensor<5xf32>
"test.reify_bound"(%2) {dim = 0}
```
Only one constraint will be added: `dim(%2, 0) = 5`. That's where the IR traversal stops (`addToWorklist = false`). Basically we add one more constraint for every op (with static dim size / const index value) where the traversal stops.

There's no easy way to test this. The information in the constraint system does not change, so the computed bounds are exactly the same. That's why I marked it NFC. This change is needed for #85895 and I put it into an extra PR to keep the other PRs smaller. (`getPos` asserts that a value/dim is mapped in the constraint set and this change makes sure that after analyzing a dim/value, it is always mapped even if it is static/const; no special handling needed in #85895.)

matthias-springer commented:
I could make this PR part of #85895. That's where this change is needed, and an assertion would be triggered in one of the test cases without this change.

matthias-springer commented:
I merged this PR into #85895. That's where it was originally and then we have a test that exercises the changes in the code.
